consensus/replay.go

```diff
@@ -315,44 +339,29 @@ func (h *Handshaker) replayBlocks(proxyApp proxy.AppConns, appBlockHeight, store
 		h.nBlocks += 1
 	}
 
-	if useReplayFunc {
+	if mutateState {
 		// sync the final block
 		return h.ReplayBlocks(appHash, finalBlock, proxyApp)
```

Comment:
This is weird, we should probably not recursively call ReplayBlocks, but rather call h.replayBlock()